router/internal/expr/expr.go

```diff
@@ -0,0 +1,169 @@
+package expr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/expr-lang/expr"
+	"github.com/expr-lang/expr/file"
+	"github.com/expr-lang/expr/vm"
+	"github.com/wundergraph/cosmo/router/pkg/authentication"
+	"net/http"
+	"net/url"
+	"reflect"
+)
+
+// RequestRootContext is the context for expressions parser when evaluating dynamic expressions
+// If possible function calls should be avoided in the expressions as they are much more expensive
+// See https://github.com/expr-lang/expr/issues/734
+type RequestRootContext struct {
+	Request Request `expr:"request"`
+}
+
+// Request is the context for the request object in expressions. Be aware, that only value receiver methods
+// are exported in the expr environment. This is because the expressions are evaluated in a read-only context.
+type Request struct {
+	Auth   RequestAuth    `expr:"auth"`
+	URL    RequestURL     `expr:"url"`
+	Header RequestHeaders `expr:"header"`
+}
+
+// RequestURL is the context for the URL object in expressions
+// it is limited in scope to the URL object and its components. For convenience, the query parameters are parsed.
+type RequestURL struct {
+	Method string `expr:"method"`
+	// Scheme is the scheme of the URL
+	Scheme string `expr:"scheme"`
+	// Host is the host of the URL
+	Host string `expr:"host"`
+	// Path is the path of the URL
+	Path string `expr:"path"`
+	// Query is the parsed query parameters
+	Query map[string]string `expr:"query"`
+}
+
+// RequestHeaders is the context for the headers object in expressions. A user can access the headers directly by
+// key and get the array representation of the header values or use the Get method to get the first value.
+type RequestHeaders map[string][]string
```

Comment:
have you considered using http.Header?

Reply:
Is this exposed by default or do we have to expose it explicitly?

Reply:
Everything is exposed by default but only exported methods are accessible within the VM.